packages/trace-viewer/src/ui/codegen.ts

```diff
@@ -0,0 +1,287 @@
+/**
+ * Copyright (c) Microsoft Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import type { Language } from '@isomorphic/locatorGenerators';
+import type * as har from '@trace/har';
+
+interface APIRequestCodegen {
+  generatePlaywrightRequestCall(request: har.Request, body: string | undefined): string;
+}
+
+class JSCodeGen implements APIRequestCodegen {
+  generatePlaywrightRequestCall(request: har.Request, body: string | undefined): string {
+    let method = request.method.toLowerCase();
+    const url = new URL(request.url);
+    const urlParam = `${url.origin}${url.pathname}`;
+    const options: any = {};
+    if (!['delete', 'get', 'head', 'post', 'put', 'patch'].includes(method)) {
+      options.method = method;
+      method = 'fetch';
+    }
+    if (url.searchParams.size)
+      options.params = Object.fromEntries(url.searchParams.entries());
+    if (body)
+      options.data = body;
+    if (request.headers.length)
+      options.headers = Object.fromEntries(request.headers.map(header => [header.name, header.value]));
+
+    const params = [`'${urlParam}'`];
+    const hasOptions = Object.keys(options).length > 0;
+    if (hasOptions)
+      params.push(this.prettyPrintObject(options));
+    return `await page.request.${method}(${params.join(', ')});`;
+  }
+
+  private prettyPrintObject(obj: any, indent = 2, level = 0): string {
+    // Handle null and undefined
+    if (obj === null)
+      return 'null';
+    if (obj === undefined)
+      return 'undefined';
+
+    // Handle primitive types
+    if (typeof obj !== 'object') {
+      if (typeof obj === 'string')
+        return `'${obj}'`;
```

Comment:
This does not look like a proper string escaping. Same problem across the languages where you only escape quotes and do not escape \\.

Reply:
good catch! fixed in https://github.com/microsoft/playwright/pull/33298/commits/a798374d1238edff59134eaa294cbf803db96abf